source/particles/particle_handler.cc

```diff
@@ -442,6 +442,204 @@ namespace Particles
 
 
 
+  template <int dim, int spacedim>
+  std::map<unsigned int, IndexSet>
+  ParticleHandler<dim, spacedim>::insert_global_particles(
+    const std::vector<Point<spacedim>> &positions,
+    const std::vector<std::vector<BoundingBox<spacedim>>>
+      &                                     global_bounding_boxes,
+    const std::vector<std::vector<double>> &properties)
+  {
+    if (!properties.empty())
+      {
+        AssertDimension(properties.size(), positions.size());
+#ifdef DEBUG
+        for (const auto &p : properties)
+          AssertDimension(p.size(), n_properties_per_particle());
+#endif
+      }
+
+    const auto tria =
+      dynamic_cast<const parallel::distributed::Triangulation<dim, spacedim> *>(
+        &(*triangulation));
+    const auto comm =
+      (tria != nullptr ? tria->get_communicator() : MPI_COMM_WORLD);
+
+    const auto n_mpi_processes = Utilities::MPI::n_mpi_processes(comm);
+
+    GridTools::Cache<dim, spacedim> cache(*triangulation, *mapping);
+
+    // Compute the global number of properties
+    const auto n_global_properties =
+      Utilities::MPI::sum(properties.size(), comm);
+
+    // Gather the number of points per processor
+    const auto n_particles_per_proc =
+      Utilities::MPI::all_gather(comm, positions.size());
+
+    // Calculate all starting points locally
+    std::vector<unsigned int> particle_start_indices(n_mpi_processes);
+
+    unsigned int particle_start_index = 0;
+    for (unsigned int process = 0; process < particle_start_indices.size();
+         ++process)
+      {
+        particle_start_indices[process] = particle_start_index;
+        particle_start_index += n_particles_per_proc[process];
+      }
+
+    // Get all local information
+    const auto cells_positions_and_index_maps =
+      GridTools::distributed_compute_point_locations(cache,
+                                                     positions,
+                                                     global_bounding_boxes);
+
+    // Unpack the information into several vectors:
+    // All cells that contain at least one particle
+    const auto &local_cells_containing_particles =
+      std::get<0>(cells_positions_and_index_maps);
+
+    // The reference position of every particle in the local part of the
+    // triangulation.
+    const auto &local_reference_positions =
+      std::get<1>(cells_positions_and_index_maps);
```

Comment:
I can't wait till we can finally have structured bindings to make the initialization of these local references simpler and easier to read...